### PR TITLE
Emit calendar interaction events

### PR DESCRIPTION
**Pull request by ChristophWurst**
This adds the calendar part of https://github.com/nextcloud/server/pull/19075.

As a follow-up I would also like to look into email attendees but for that I need to parse the calendar data and I'm unsure how to do this reliably.

### CODE REVIEW

**Review by julien-nc — COMMENTED** [review]
Nice!
What about when a calendar object is deleted? Do you consider this is not an interaction or is it just not possible to get that event?

**Comment by ChristophWurst** [comment]
> What about when a calendar object is deleted? Do you consider this is not an interaction or is it just not possible to get that event?

I see your point but I'm not sure about how much this makes sense from a user PoV. Deleting an event or removing a share doesn't seems like a real interaction with someone else. Technically this shouldn't be a barrier to add though. :shrug:

**Comment by eneiluj** [comment]
> I see your point but I'm not sure about how much this makes sense from a user PoV. Deleting an event or removing a share doesn't seems like a real interaction with someone else. Technically this shouldn't be a barrier to add though. shrug

Right. Anyway the delete event is available for other use cases.

**Review by julien-nc — APPROVED** [review]

File: apps/dav/lib/Listener/CalendarContactInteractionListener.php [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+		if ($event instanceof CalendarShareUpdatedEvent && !empty($event->getAdded())) {
+			// group: href => principal:principals/groups/admin
+			// users: href => principal:principals/users/admin
+			foreach ($event->getAdded() as $added) {
+				if (!isset($added['href'])) {
+					// Nothing to work with
+					continue;
+				}
+				$this->emitFromUri($added, $user);
```
Comment (kesselb) [inline apps/dav/lib/Listener/CalendarContactInteractionListener.php:93]:
```suggestion
				$this->emitFromUri($added['href'], $user);
```

I'm not sure. But above isset checks for href.

**Comment by rullzer** [comment]
Master is Nextcloud 22 now.
If this should go into 21 it should be backported.

**Comment by tcitworld** [comment]
> As a follow-up I would also like to look into email attendees but for that I need to parse the calendar data and I'm unsure how to do this reliably.

Something like:
```php
VObject\Reader::read($calendarData, VObject\Reader::OPTION_FORGIVING);
```
Then something like this: https://github.com/nextcloud/server/blob/17930c64f499979354d68e80e91c5ad3536d0c84/apps/dav/lib/CalDAV/Reminder/NotificationProvider/EmailProvider.php#L247

File: apps/dav/lib/Listener/CalendarContactInteractionListener.php (hunk 1 above)
Comment (ChristophWurst) [reply apps/dav/lib/Listener/CalendarContactInteractionListener.php]: fixed!

**Comment by ChristophWurst** [comment]
@tcitworld thanks so much! I will do this in a follow-up nevertheless because none of those methods are reusable and I'm not sure if I want to copy the logic or restructure to make it reusable.

**Review by kesselb — APPROVED** [review]

**Pull request merged**